src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/persistence/PublisherEventsPersistor.java

```diff
@@ -0,0 +1,80 @@
+
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.PerformanceAnalyzerApp;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.decisionmaker.actions.Action;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.metrics.ExceptionsAndErrors;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.metrics.RcaRuntimeMetrics;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.persistence.actions.PersistedAction;
+import java.time.Instant;
+import java.util.stream.Collectors;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * A listener that persists all actions published by the publisher to rca.sqlite
+ */
+public class PublisherEventsPersistor {
+    public static final String NAME = "publisher_events_persistor";
+    private static final Logger LOG = LogManager.getLogger(PublisherEventsPersistor.class);
+
+    private final Persistable persistable;
+
+    public PublisherEventsPersistor(final Persistable persistable) {
+        this.persistable = persistable;
+    }
+
+    public void persistAction(final Action action) {
+        long timestamp = Instant.now().toEpochMilli();
+        if (persistable != null) {
+            LOG.debug("Action: [{}] published to persistor publisher.", action.name());
+            PerformanceAnalyzerApp.RCA_RUNTIME_METRICS_AGGREGATOR.updateStat(
+                    RcaRuntimeMetrics.ACTIONS_PUBLISHED, action.name(), 1);
+            if (action.impactedNodes() != null) {
+                final String nodeIds = action.impactedNodes().stream()
+                                        .map(n -> n.getNodeId().toString())
+                                        .collect(Collectors.joining(",", "{", "}"));
+                final String nodeIps = action.impactedNodes().stream()
+                                         .map(n -> n.getHostAddress().toString())
+                                         .collect(Collectors.joining(",", "{", "}"));
+                final PersistedAction actionsSummary = new PersistedAction();
+                actionsSummary.setActionName(action.name());
+                actionsSummary.setNodeIds(nodeIds);
+                actionsSummary.setNodeIps(nodeIps);
+                actionsSummary.setActionable(action.isActionable());
+                actionsSummary.setCoolOffPeriod(action.coolOffPeriodInMillis());
+                actionsSummary.setMuted(action.isMuted());
+                actionsSummary.setSummary(action.summary());
+                actionsSummary.setTimestamp(timestamp);
+                try {
+                    persistable.write(actionsSummary);
+                } catch (Exception e) {
+                    LOG.error("Unable to write publisher events to sqlite", e);
+                    PerformanceAnalyzerApp.ERRORS_AND_EXCEPTIONS_AGGREGATOR.updateStat(
+                            ExceptionsAndErrors.EXCEPTION_IN_PERSIST, this.name(), 1);
```

Comment:
Maybe this.name() is a redundant dimension for this metric as it does not change from action to action. Do we want to capture the name of the Action we wanted to persist ?

Reply:
Right. Modified.